File: scenarios/basic/desktop/linux-qt/MissingImageAltText.py
import sys
from PyQt5.QtWidgets import (QApplication, QMainWindow, QVBoxLayout, QHBoxLayout, 
                             QWidget, QLabel, QPushButton, QFrame, QGridLayout, QScrollArea)
from PyQt5.QtCore import Qt
from PyQt5.QtGui import QFont, QPixmap

class MissingImageAltText(QMainWindow):
    def __init__(self):
        super().__init__()
        self.setWindowTitle("TechStore")
        self.setGeometry(100, 100, 1200, 800)
        
        # Central widget
        central_widget = QWidget()
        self.setCentralWidget(central_widget)
        
        # Main layout
        main_layout = QVBoxLayout(central_widget)
        main_layout.setSpacing(20)
        
        # Header
        header_layout = QVBoxLayout()
        header_layout.setAlignment(Qt.AlignCenter)
        
        title_label = QLabel("TechStore")
        title_font = QFont()
        title_font.setPointSize(24)
        title_font.setBold(True)
        title_label.setFont(title_font)
        title_label.setAlignment(Qt.AlignCenter)
        header_layout.addWidget(title_label)
        
        subtitle_label = QLabel("Discover amazing tech products")
        subtitle_font = QFont()
        subtitle_font.setPointSize(16)
        subtitle_label.setFont(subtitle_font)
        subtitle_label.setAlignment(Qt.AlignCenter)
        header_layout.addWidget(subtitle_label)
        
        main_layout.addLayout(header_layout)
        
        # Category filter
        category_layout = QHBoxLayout()
        category_layout.setAlignment(Qt.AlignCenter)
        category_layout.setSpacing(10)
        
        categories = ["All", "Electronics", "Clothing", "Home", "Sports"]
        for category in categories:
            button = QPushButton(category)
            if category == "All":
                button.setStyleSheet("QPushButton { background-color: #007bff; color: white; font-weight: bold; }")
            button.setFixedSize(100, 40)
            category_layout.addWidget(button)
        
        main_layout.addLayout(category_layout)
        
        # Products scrollable area
        scroll_area = QScrollArea()
        scroll_area.setWidgetResizable(True)
        scroll_area.setVerticalScrollBarPolicy(Qt.ScrollBarAsNeeded)
        scroll_area.setHorizontalScrollBarPolicy(Qt.ScrollBarAsNeeded)
        
        # Products container
        products_widget = QWidget()
        products_layout = QGridLayout(products_widget)
        products_layout.setSpacing(20)
        
        # Load products
        self.load_products(products_layout)
        
        scroll_area.setWidget(products_widget)
        main_layout.addWidget(scroll_area)
        
    def load_products(self, layout):
        products = [
            {
                "name": "Wireless Headphones",
                "price": "$99.99",
                "original_price": "$149.99",
                "rating": "4.5",
                "image": "🎧"
            },
            {
                "name": "Smart Watch",
                "price": "$299.99",
                "original_price": "$399.99",
                "rating": "4.8",
                "image": "⌚"
            },
            {
                "name": "Portable Speaker",
                "price": "$79.99",
                "original_price": "$99.99",
                "rating": "4.3",
                "image": "🔊"
            },
            {
                "name": "Gaming Mouse",
                "price": "$59.99",
                "original_price": "$79.99",
                "rating": "4.6",
                "image": "🖱️"
            },
            {
                "name": "Mechanical Keyboard",
                "price": "$129.99",
                "original_price": "$179.99",
                "rating": "4.7",
                "image": "⌨️"
            },
            {
                "name": "USB-C Hub",
                "price": "$49.99",
                "original_price": "$69.99",
                "rating": "4.2",
                "image": "🔌"
            }
        ]
        
        row = 0
        col = 0
        max_cols = 3
        
        for product in products:
            product_frame = self.create_product_card(product)
            layout.addWidget(product_frame, row, col)
            
            col += 1
            if col >= max_cols:
                col = 0
                row += 1
    
    def create_product_card(self, product):
        # Product frame
        frame = QFrame()
        frame.setFrameStyle(QFrame.Box)
        frame.setFixedSize(300, 400)
        frame.setStyleSheet("QFrame { background-color: white; border: 1px solid #ddd; }")
        
        layout = QVBoxLayout(frame)
        layout.setSpacing(10)
        layout.setContentsMargins(10, 10, 10, 10)
        
        # MISSING IMAGE ALT TEXT - Product image without alternative text
        image_label = QLabel(product["image"])
        image_label.setAlignment(Qt.AlignCenter)
        image_label.setStyleSheet("font-size: 100px;")
        image_label.setFixedHeight(200)
        layout.addWidget(image_label)
        
        # Product name
        name_label = QLabel(product["name"])
        name_font = QFont()
        name_font.setPointSize(16)
        name_font.setBold(True)
        name_label.setFont(name_font)
        name_label.setWordWrap(True)
        layout.addWidget(name_label)
        
        # Rating
        rating_label = QLabel(f"⭐ {product['rating']}")
        layout.addWidget(rating_label)
        
        # Price
        price_layout = QHBoxLayout()
        
        current_price = QLabel(product["price"])
        price_font = QFont()
        price_font.setPointSize(16)
        price_font.setBold(True)
        current_price.setFont(price_font)
        current_price.setStyleSheet("color: green;")
        price_layout.addWidget(current_price)
        
        original_price = QLabel(product["original_price"])
        original_price.setStyleSheet("text-decoration: line-through; color: gray;")
        price_layout.addWidget(original_price)
        
        price_layout.addStretch()
        layout.addLayout(price_layout)
        
        # Add to cart button
        add_to_cart_button = QPushButton("Add to Cart")
        add_to_cart_button.setStyleSheet("QPushButton { background-color: #28a745; color: white; font-weight: bold; }")
        add_to_cart_button.setFixedHeight(40)
        layout.addWidget(add_to_cart_button)
        
        return frame

def main():
    app = QApplication(sys.argv)
    window = MissingImageAltText()
    window.show()
    sys.exit(app.exec_())

if __name__ == "__main__":
    main()

File: scenarios/basic/cross-platform/qt/MissingImageAltText.cpp
#include <QApplication>
#include <QMainWindow>
#include <QVBoxLayout>
#include <QHBoxLayout>
#include <QGridLayout>
#include <QLabel>
#include <QPushButton>
#include <QWidget>
#include <QScrollArea>
#include <QFrame>

class ECommerceCatalog : public QMainWindow
{
    Q_OBJECT

public:
    ECommerceCatalog(QWidget *parent = nullptr) : QMainWindow(parent)
    {
        setupUI();
        setupConnections();
    }

private slots:
    void addToCart()
    {
        // Simulate adding item to cart
        QMessageBox::information(this, "Cart", "Item added to cart!");
    }
    
    void viewDetails()
    {
        QMessageBox::information(this, "Details", "Viewing product details...");
    }
    
    void addToWishlist()
    {
        QMessageBox::information(this, "Wishlist", "Item added to wishlist!");
    }

private:
    void setupUI()
    {
        QWidget *centralWidget = new QWidget(this);
        setCentralWidget(centralWidget);
        
        QVBoxLayout *mainLayout = new QVBoxLayout(centralWidget);
        
        // Header
        QHBoxLayout *headerLayout = new QHBoxLayout();
        
        QLabel *titleLabel = new QLabel("Electronics Store", this);
        titleLabel->setStyleSheet(
            "QLabel {"
            "    font-size: 28px;"
            "    font-weight: bold;"
            "    color: #333;"
            "    padding: 20px 0;"
            "}"
        );
        headerLayout->addWidget(titleLabel);
        
        headerLayout->addStretch();
        
        QPushButton *cartBtn = new QPushButton("🛒 Cart (0)", this);
        cartBtn->setStyleSheet(
            "QPushButton {"
            "    background: #007bff;"
            "    color: white;"
            "    border: none;"
            "    border-radius: 6px;"
            "    padding: 10px 20px;"
            "    font-size: 16px;"
            "    font-weight: bold;"
            "}"
            "QPushButton:hover {"
            "    background: #0056b3;"
            "}"
        );
        headerLayout->addWidget(cartBtn);
        
        mainLayout->addLayout(headerLayout);
        
        // Product grid
        QScrollArea *scrollArea = new QScrollArea(this);
        scrollArea->setWidgetResizable(true);
        scrollArea->setHorizontalScrollBarPolicy(Qt::ScrollBarAlwaysOff);
        
        QWidget *scrollWidget = new QWidget();
        QGridLayout *gridLayout = new QGridLayout(scrollWidget);
        gridLayout->setSpacing(20);
        
        // Product 1
        QFrame *product1 = createProductCard(
            "Smartphone Pro Max",
            "$999.99",
            "Latest smartphone with advanced features",
            "📱"
        );
        gridLayout->addWidget(product1, 0, 0);
        
        // Product 2
        QFrame *product2 = createProductCard(
            "Wireless Headphones",
            "$299.99",
            "Premium noise-canceling headphones",
            "🎧"
        );
        gridLayout->addWidget(product2, 0, 1);
        
        // Product 3
        QFrame *product3 = createProductCard(
            "Gaming Laptop",
            "$1,599.99",
            "High-performance gaming laptop",
            "💻"
        );
        gridLayout->addWidget(product3, 0, 2);
        
        // Product 4
        QFrame *product4 = createProductCard(
            "Smart Watch",
            "$399.99",
            "Fitness tracking smartwatch",
            "⌚"
        );
        gridLayout->addWidget(product4, 1, 0);
        
        // Product 5
        QFrame *product5 = createProductCard(
            "Tablet Pro",
            "$799.99",
            "Professional tablet for work and creativity",
            "📱"
        );
        gridLayout->addWidget(product5, 1, 1);
        
        // Product 6
        QFrame *product6 = createProductCard(
            "Bluetooth Speaker",
            "$149.99",
            "Portable wireless speaker with great sound",
            "🔊"
        );
        gridLayout->addWidget(product6, 1, 2);
        
        scrollArea->setWidget(scrollWidget);
        mainLayout->addWidget(scrollArea);
    }
    
    QFrame* createProductCard(const QString &name, const QString &price, const QString &description, const QString &icon)
    {
        QFrame *card = new QFrame(this);
        card->setStyleSheet(
            "QFrame {"
            "    background: white;"
            "    border: 1px solid #ddd;"
            "    border-radius: 8px;"
            "    padding: 20px;"
            "}"
            "QFrame:hover {"
            "    box-shadow: 0 4px 12px rgba(0,0,0,0.1);"
            "    transform: translateY(-2px);"
            "}"
        );
        
        QVBoxLayout *cardLayout = new QVBoxLayout(card);
        cardLayout->setSpacing(15);
        
        // MISSING IMAGE ALT TEXT - Product image without alternative text
        QLabel *productImage = new QLabel(this);
        productImage->setMinimumHeight(200);
        productImage->setStyleSheet(
            "QLabel {"
            "    background: qlineargradient(x1:0, y1:0, x2:1, y2:1, stop:0 #f8f9fa, stop:1 #e9ecef);"
            "    border: 2px dashed #dee2e6;"
            "    border-radius: 6px;"
            "    font-size: 48px;"
            "    color: #6c757d;"
            "}"
        );
        productImage->setText(icon);
        productImage->setAlignment(Qt::AlignCenter);
        cardLayout->addWidget(productImage);
        
        // Product name
        QLabel *nameLabel = new QLabel(name, this);
        nameLabel->setStyleSheet(
            "QLabel {"
            "    font-size: 18px;"
            "    font-weight: bold;"
            "    color: #333;"
            "    margin-bottom: 5px;"
            "}"
        );
        nameLabel->setWordWrap(true);
        cardLayout->addWidget(nameLabel);
        
        // Product description
        QLabel *descLabel = new QLabel(description, this);
        descLabel->setStyleSheet(
            "QLabel {"
            "    font-size: 14px;"
            "    color: #666;"
            "    margin-bottom: 10px;"
            "}"
        );
        descLabel->setWordWrap(true);
        cardLayout->addWidget(descLabel);
        
        // Price
        QLabel *priceLabel = new QLabel(price, this);
        priceLabel->setStyleSheet(
            "QLabel {"
            "    font-size: 20px;"
            "    font-weight: bold;"
            "    color: #007bff;"
            "    margin-bottom: 15px;"
            "}"
        );
        cardLayout->addWidget(priceLabel);
        
        // Action buttons
        QHBoxLayout *buttonLayout = new QHBoxLayout();
        
        QPushButton *addToCartBtn = new QPushButton("Add to Cart", this);
        addToCartBtn->setStyleSheet(
            "QPushButton {"
            "    background: #28a745;"
            "    color: white;"
            "    border: none;"
            "    border-radius: 4px;"
            "    padding: 8px 16px;"
            "    font-size: 14px;"
            "    font-weight: bold;"
            "}"
            "QPushButton:hover {"
            "    background: #218838;"
            "}"
        );
        connect(addToCartBtn, &QPushButton::clicked, this, &ECommerceCatalog::addToCart);
        buttonLayout->addWidget(addToCartBtn);
        
        QPushButton *detailsBtn = new QPushButton("Details", this);
        detailsBtn->setStyleSheet(
            "QPushButton {"
            "    background: #6c757d;"
            "    color: white;"
            "    border: none;"
            "    border-radius: 4px;"
            "    padding: 8px 16px;"
            "    font-size: 14px;"
            "    font-weight: bold;"
            "}"
            "QPushButton:hover {"
            "    background: #5a6268;"
            "}"
        );
        connect(detailsBtn, &QPushButton::clicked, this, &ECommerceCatalog::viewDetails);
        buttonLayout->addWidget(detailsBtn);
        
        QPushButton *wishlistBtn = new QPushButton("♡", this);
        wishlistBtn->setStyleSheet(
            "QPushButton {"
            "    background: #dc3545;"
            "    color: white;"
            "    border: none;"
            "    border-radius: 4px;"
            "    padding: 8px 12px;"
            "    font-size: 16px;"
            "    min-width: 40px;"
            "}"
            "QPushButton:hover {"
            "    background: #c82333;"
            "}"
        );
        connect(wishlistBtn, &QPushButton::clicked, this, &ECommerceCatalog::addToWishlist);
        buttonLayout->addWidget(wishlistBtn);
        
        cardLayout->addLayout(buttonLayout);
        
        return card;
    }
    
    void setupConnections()
    {
        // Additional connections can be added here
    }
};

int main(int argc, char *argv[])
{
    QApplication app(argc, argv);
    
    ECommerceCatalog catalog;
    catalog.setWindowTitle("Electronics Store - Product Catalog");
    catalog.resize(1200, 800);
    catalog.show();
    
    return app.exec();
}

#include "MissingImageAltText.moc"
